### PR TITLE
[GSoC 2018] LIMS Integration into LibreEHR

### PR DESCRIPTION
**Pull request by Perseus**
Initial commit for LIMS. 
Contains basic directory structure, the first draft of installation instructions for SENAITE, and global settings for all things required for LIMS integration.

@teryhill Please add the tags whenever possible.

### CODE REVIEW

**Comment by aethelwulffe** [comment]
Please accept collaboration invitation to this repo.

-I recommend hijacking, removing, relocating, or refactoring anything related to "labs" existing in the system at some point in the project.  It would be nice if this effort sort of took ownership of all the legacy mess.
Beautiful Start.  Do we want to do this in staged integration, or mark this as "WIP" and keep the branch alive?

File: modules/LIMS/senaite/README.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+* Microsoft Windows
+* LINUX distributions
+* OS X (Instructions will be added soon)
+
+## Microsoft Windows
+
+To install SENAITE on Windows, you will need the following resources:
+
+* [VirtualBox](https://www.virtualbox.org/)
```
[the remark this answers is not in the document]
Comment (teryhill) [reply modules/LIMS/senaite/README.md]: That would be something that @muarachmann  needs to know about for the installer.

File: library/globals.inc.php [hunk 2]
```diff
@@ -3217,10 +3224,48 @@
...
+        xl('Address where the LIMS backend API is available')
+      ),
+      'lims_username' => array(
+        xl('LIMS Authentication Username'),
+        'text',
+        '',
+        xl('Username provided by the LIMS for login/API purposes')
+      ),
+      'lims_password' => array(
+        xl('LIMS Authentication Password'),
```
[the remark this answers is not in the document]
Comment (teryhill) [reply library/globals.inc.php]: It the 3rd party uses password the I think it should be used here as to not confuse the user.

File: library/globals.inc.php [hunk 3]
```diff
@@ -3217,10 +3224,48 @@
...
+        'text',
```
[the remark this answers is not in the document]
Comment (aethelwulffe) [reply library/globals.inc.php]: Agreed/.

File: modules/LIMS/senaite/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (aethelwulffe) [reply modules/LIMS/senaite/README.md]: Right.

File: library/globals.inc.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (aethelwulffe) [reply library/globals.inc.php]: I'm sure you are right, but of course if someone is doing  linux VM install, it probably won't confuse them much 😆

File: library/globals.inc.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Perseus) [reply library/globals.inc.php]: Ahh I tried password and it didn't work so I thought there wasn't an option for passwords. Will do the change.

File: modules/LIMS/senaite/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Perseus) [reply modules/LIMS/senaite/README.md]: Will look into creating a pre-configured VM. Issue with that would be getting the Plone CMS Authentication details as those are displayed on the terminal post installation so I'll have to look for a way to have them stored in a file or something while installing

[the remark this answers is not in the document]
Comment (Perseus) [reply modules/LIMS/senaite/README.md]: Think I've figured out how to create a pre-configured ISO file containing an installation of plone+senaite. What d'you think I should do with the pass given by Plone and also the password of the OS? I can make a normal password and store it in a text file on the desktop and let them know that they should change it to something of their preference ( could do this in the readme as well )

[the remark this answers is not in the document]
Comment (teryhill) [reply modules/LIMS/senaite/README.md]: That will be a good approach. We request the users change passwords currently and the current  default passwords are in the installation.

**Comment by teryhill** [comment]
@Perseus I tested this using @naveen17797 's updater in developer mode and the changes are there to globals. Looking forward to the next set of changes. Did your question about where to put the ISO for this get answered?

**Comment by aethelwulffe** [comment]
Oh.  Location for the ISO.  Should we discuss now?  
  Footprint?  May determine where we can put it.
  Need a temporary host?  I can provide.  MI2 may be able to as well, if LH cannot.  They had no available and maintainable file server on which to put the prepackaged xammp distros of LibreHealth either, so we are not keeping a current copy in distribution.  I imagine they cannot support hosting an ISO either.  We should formalize a location.  I don't believe GitHub supports files over 100 meg.
  What is your preference?

**Comment by Perseus** [comment]
File size is approx 1.8GB. I was considering Google drive since data seems to stay on it for a long time but I'm unaware of any long-term problems that it may cause.

**Comment by aethelwulffe** [comment]
If a googlie thing is doable, and the download does not require any token from Google, then that is a potential mirror.

**Comment by aethelwulffe** [comment]
We can always stick it on the SCC HoneyPot server.

**Comment by Perseus** [comment]
Here's a link to the google drive upload
https://drive.google.com/open?id=1X8r9Zc5zW4p6-8BHQMDrQVG0ELDh3VEK
No token is required, link is shareable

**Comment by aethelwulffe** [comment]
Cool.  We should also start looking at https://www.tuxfamily.org/en/about too.

**Comment by Perseus** [comment]
TuxFamily looks good. They have a 1GB limit on the downloads but they say they can increase it if needed

**Comment by aethelwulffe** [comment]
There might be the possibility of paring down the ISO too.  Would it not be hilarious to build something on Minix?

**Comment by Perseus** [comment]
@teryhill @aethelwulffe I was thinking about the LIMS password storage from a security point of view and it probably wouldn't be too secure having the password stored in our database in cleartext like that. Maybe we can just store the username and let them type the password to log into the LIMS? I thought about storing the password in a hash but since I wouldn't be able to unhash it, the data won't be proper while being sent to the LIMS for authentication. Thoughts?

**Comment by Perseus** [comment]
Couple screenshots to go with these commits:
image
image
image
image

I plan on redesigning the dashboard a bit after all the rest of the features are finished. Order of features differs from my proposal a bit since different components had different prerequisites which needed to be done first.

**Comment by teryhill** [comment]
@Perseus can you do a rebase and try to clean up the PR. It has a number of old pr's in it and you also have a conflict with edit_globals.php .

**Comment by Perseus** [comment]
Sorry @teryhill my next push should have that fixed. I had some issues with git and windows not differentiating between branch names that had case differences, so I ended up creating a different branch (lims instead of LIMS) which messed some previous commits up

**Comment by teryhill** [comment]
Not a problem @Perseus . Do you have the setup worked out and documented?

**Comment by Perseus** [comment]
Almost. The base setup works, I have to write some more documentation and add a small line script for user permissions on the API folder. I will push another commit with updates tomorrow or day after at the latest. That should have the updated docs (both dev and consumer).

**Comment by aethelwulffe** [comment]
LIMS vs lims on windows 10:
For directories with spaces, use double quotes like so:
fsutil.exe file setCaseSensitiveInfo "C:\my repos\LIMS" enable
Without spaces:
fsutil.exe file setCaseSensitiveInfo C:\myrepos\LIMS enable

**Comment by aethelwulffe** [comment]
git config core.ignorecase false
The above tells GIT itself to use case sensitivity...Sorry, this should be the answer for branch names, not the above.

**Comment by aethelwulffe** [comment]
I guess it looks like most cURL library projects out there use composer now.  The only one I have ever used is https://github.com/shuber/curl , and that was a Looooong time ago. In the end, I discovered that I really didn't need the wrapper just to get the POST and cookie anyway, but that project at least taught me how to do it.  Took about 5 lines of code in the end.  I don't know how secure it was or anything, but the basic use of cURL was simple.  The HARD part of using cURL is the incredibly shitty documentation, and the total lack of indication as to what commands are using what libraries.  There was also great confusion on my part as to how to enable cURL in the first place (long time ago...had only just entered to nonsensical world of web scripting a couple of years before).

I suppose package managers are great for some folks.  It just always seems to me that they result in tons of unneeded and slow deadweight.  I am sure they are great for building out single-unit programs, but when you have a giant pile of applications all rolled into one, you wind up with...a pile of stuff.  It's just too bad that the best you can do is to include that pile as an artifact restricted to the composer directory structure.

-  Yeah, we could look at using this in /assets in the future, but I guess it is best to leave it where it is.  I could see that it was a composer project, but I didn't realize that most of the content was related to one library being a composer project.  

On topic:  As I said, review at this point is a little hard, so could you tell me if this wrapper or your code gives a useful error (from a liberal arts major point of view) if cURL is not enabled on the server?

**Comment by Perseus** [comment]
I collaborated with @muarachmann and requested him to put in a requirement for cURL in the setup with a small notification that it's for use with LIMS. I haven't, yet, put in any error message inside the LIMS implementation itself concerning the status of cURL on the server, I'll put that in now.

**Comment by aethelwulffe** [comment]
Why did this branch get closed?

**Comment by Perseus** [comment]
@aethelwulffe I had some rebase/merge issues with my branch so I created a new one. I'll make a PR with it tomorrow

**Pull request closed without merging**